[Automation] native-image run fails for io.netty:netty-codec-smtp:4.2.1.Final

## native-image run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: native-image run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25251908980/job/74045252632
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
erated/handlers/reflect-config.json+api=META-INF/native-image/io.netty/netty-codec-smtp/generated/handlers/reflect-config.json' \
'-H:ReflectionConfigurationResources@jar:file:///home/runner/.gradle/caches/modules-2/files-2.1/io.netty/netty-codec-base/4.2.1.Final/ee2a21f30549d6fb0ec1da71ac748c385368d807/netty-codec-base-4.2.1.Final.jar!/META-INF/native-image/io.netty/netty-codec-base/generated/handlers/reflect-config.json+api=META-INF/native-image/io.netty/netty-codec-base/generated/handlers/reflect-config.json' \
-H:FallbackThreshold@user+api=0 \
-H:Name@user+api=io.netty.netty-codec-smtp_tests-tests \
-H:ConfigurationFileDirectories@user=/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/generated/generateTestResourcesConfigFile,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.opentest4j/opentest4j/1.2.0,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.junit.platform/junit-platform-commons/1.8.2,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.hamcrest/hamcrest-core/1.3,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-codec-smtp/4.1.74.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-common/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-buffer/4.1.80.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-transport/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-resolver/4.1.79.Final \
-H:Optimize@user+api=b \
-H:+ReportExceptionStackTraces@user \
-H:Features@user+api=org.graalvm.junit.platform.JUnitPlatformFeature \
'-H:Class@explicit main-class=org.graalvm.junit.platform.NativeImageJUnitLauncher' \
-H:ImageBuildID@driver=31079edf-2d4a-6bcc-23b8-0f4af382112d \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-25_linux-x64_bin/25.0.0/x64/graalvm-jdk-25.0.3+9.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/thirdparty/native-image.properties+api=com.oracle.svm.thirdparty.gson.GsonFeature' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-25_linux-x64_bin/25.0.0/x64/graalvm-jdk-25.0.3+9.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature,com.oracle.svm.polyglot.scala.ScalaFeature'
]
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-smtp_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                    (6.0s @ 0.14GB)
 Java version: 25.0.3+9-LTS, vendor version: Oracle GraalVM 25.0.3+9.1
 Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 C compiler: gcc (linux, x86_64, 13.3.0)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 2 user-specific feature(s):
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - org.graalvm.junit.platform.JUnitPlatformFeature
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 12.67GB of memory (75.6% of system memory, $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [*****]                                                                   (33.9s @ 0.90GB)
    8,908 types,  10,132 fields, and  39,569 methods found reachable
    3,181 types,     728 fields, and   4,037 methods registered for reflection
       62 types,      68 fields, and      53 methods registered for JNI access
        0 downcalls and 0 upcalls registered for foreign access
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                               (4.1s @ 1.01GB)
[4/8] Parsing methods...      [**]                                                                       (4.0s @ 0.66GB)
[5/8] Inlining methods...     [***]                                                                      (2.2s @ 0.71GB)
[6/8] Compiling methods...    [****]                                                                    (18.5s @ 0.87GB)
[7/8] Laying out methods...   [**]                                                                       (3.7s @ 0.83GB)
[8/8] Creating image...       [**]                                                                       (3.5s @ 0.95GB)
  10.14MB (35.84%) for code area:    24,630 compilation units
  14.09MB (49.80%) for image heap:  221,287 objects and 151 resources
   4.06MB (14.36%) for other data
  28.30MB in total image size, 26.09MB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
   4.63MB java.base                                            2.99MB byte[] for code metadata
   2.19MB svm.jar (Native Image)                               2.72MB byte[] for java.lang.String
 597.63kB assertj-core-3.22.0.jar                              1.43MB java.lang.String
 481.34kB java.xml                                             1.12MB java.lang.Class
 262.19kB netty-common-4.2.1.Final.jar                       926.43kB com.oracle.svm.core.hub.DynamicHubCompanion
 251.02kB netty-buffer-4.2.1.Final.jar                       708.54kB byte[] for general heap data
 198.00kB junit-jupiter-engine-5.9.2.jar                     451.55kB java.util.concurrent.ConcurrentHashMap$Node
 124.47kB netty-transport-4.2.1.Final.jar                    390.05kB heap alignment
 120.81kB junit-platform-launcher-1.9.2.jar                  359.26kB byte[] for reflection metadata
 103.55kB java.logging                                       301.15kB java.util.HashMap$Node
 793.71kB for 31 more packages                                 2.69MB for 2053 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 22 component(s) is embedded in binary (1.07kB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 FUTR: Use '--future-defaults=all' to prepare for future releases.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                        5.6s (7.1% of total time) in 802 GCs | Peak RSS: 1.75GB | CPU load: 3.60
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests (executable)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library)
 /tmp/native-image-build-output.json (build_info)
========================================================================================================================
Finished generating 'io.netty.netty-codec-smtp_tests-tests' in 1m 18s.
[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests'' finished with non-zero exit value 1

* Try:
> > Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpCommandValueOfReusesKnownCommandsAndKeepsCustomNames() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderStreamsDataContentAndTerminatesLastContent() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpContentPreservesConcreteTypesAcrossCopyDuplicateAndReplace() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderRejectsLinesLongerThanConfiguredMaximum() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderWaitsForCompleteLinesAndRejectsMalformedLines() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderAccumulatesMultiLineResponsesAcrossInboundWrites() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > defaultMessagesValidateCodesAndImplementValueSemantics() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestFactoryCreatesProtocolSpecificCommandsAndImmutableParameters() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderParsesSingleLineAndMultiLineResponses() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines() FAILED


Failures (3):
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderStreamsDataContentAndTerminatesLastContent()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderStreamsDataContentAndTerminatesLastContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 6 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[        11 tests found           ]
[         0 tests skipped         ]
[        11 tests started         ]
[         0 tests aborted         ]
[         8 tests successful      ]
[         3 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date
Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 21s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-smtp:4.1.74.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 23s
11 actionable tasks: 3 executed, 8 up-to-date
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"]
```

## native-image run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: native-image run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25251908980/job/74045252627
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
.platform/junit-platform-commons/1.8.2,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.hamcrest/hamcrest-core/1.3,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-codec-smtp/4.1.74.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-common/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-buffer/4.1.80.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-transport/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-resolver/4.1.79.Final \
-H:Optimize@user+api=b \
-H:+ReportExceptionStackTraces@user \
-H:Features@user+api=org.graalvm.junit.platform.JUnitPlatformFeature \
'-H:Class@explicit main-class=org.graalvm.junit.platform.NativeImageJUnitLauncher' \
-H:ImageBuildID@driver=31079edf-2d4a-6bcc-23b8-0f4af382112d \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/thirdparty/native-image.properties+api=com.oracle.svm.thirdparty.gson.GsonFeature' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature' \
'-H:AddExports@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=java.base/jdk.internal.misc=org.graalvm.nativeimage.guest' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=com.oracle.svm.guest.features.jdk.hosted.JavaNetHttpFeature' \
'-H:ClassInitialization@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=sun.net.httpserver.simpleserver:run_time' \
'-H:ReachabilityMetadataResources@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json+api=META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json' \
-H:DriverWarningsCount=4
]
Warning: Option 'FallbackThreshold' is deprecated and might be removed in a future release: It no longer has any effect, and no replacement is available. Please refer to the GraalVM release notes.
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-smtp_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                   (7.4s @ 0.21GiB)
 Builder configuration:
 - Java version: 25.0.2+10-LTS, vendor version: Oracle GraalVM 25.1.0-dev+10.1
 - Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 - C compiler: gcc (linux, x86_64, 13.3.0)
 - Assertions: disabled, system assertions: disabled
 - 1 user-specific feature(s):
   - org.graalvm.junit.platform.JUnitPlatformFeature
 Image configuration:
 - Garbage collector: Serial GC (max heap size: 80% of RAM)
 - Assertions: disabled (class-specific config may apply), system assertions: disabled
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 13.27GiB of memory (85% of system memory because $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [*****]                                                                  (37.5s @ 0.86GiB)
    8,698 types,   9,875 fields, and  38,889 methods found reachable
      351 types,     118 fields, and     766 methods registered for reflection
       60 types,      68 fields, and      52 methods registered for JNI access
      151 resource accesses registered with 262.18KiB total size
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                              (4.5s @ 0.96GiB)
[4/8] Parsing methods...      [**]                                                                      (4.4s @ 0.64GiB)
[5/8] Inlining methods...     [***]                                                                     (2.1s @ 0.68GiB)
[6/8] Compiling methods...    [*****]                                                                  (21.9s @ 0.70GiB)
[7/8] Laying out methods...   [**]                                                                      (4.0s @ 0.86GiB)
[8/8] Creating image...       [**]                                                                      (3.7s @ 0.98GiB)
   8.74MiB (33.70%) for code area:    24,132 compilation units
  13.38MiB (51.55%) for image heap:  221,213 objects and 21 resources
   3.83MiB (14.75%) for other data
  25.95MiB in total image size, 23.88MiB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
 993.54KiB o.g.n.runtime/com.oracle.svm.core.code              2.73MiB byte[] for code metadata
 700.60KiB java.base/java.util                                 2.62MiB byte[] for string data
 408.45KiB java.base/jdk.internal.classfile.impl               1.40MiB java.base/java.lang.String
 305.84KiB java.base/java.lang                                 1.04MiB java.base/java.lang.Class
 230.63KiB io.netty.buffer                                   951.34KiB o.g.n.runtime/c.o.s.core.hub.DynamicHubCompanion
 228.82KiB java.base/java.text                               578.52KiB byte[] for general heap data
 226.63KiB java.base/java.util.concurrent                    493.72KiB byte[] for reflection metadata
 224.55KiB java.base/com.sun.crypto.provider                 337.15KiB o.g.n.runtime/c.o.s.c.h.ImageReflectionMetadata
 223.93KiB org.assertj.core.internal.bytebuddy.jar.asm       298.81KiB java.base/java.util.HashMap$Node
 164.05KiB java.base/java.lang.invoke                        262.52KiB byte[] for embedded resources
   5.12MiB for 326 more packages                               2.73MiB for 2051 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 22 component(s) is embedded in binary (1.13KiB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 FUTR: Use '--future-defaults=all' to prepare for future releases.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                       5.8s (6.5% of total time) in 757 GCs | Peak RSS: 1.61GiB | CPU load: 3.58
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests (executable, 23.88MiB)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library, 32.23KiB)
 /tmp/native-image-build-output.json (build_info, 1.03KiB)
========================================================================================================================
Finished generating 'io.netty.netty-codec-smtp_tests-tests' in 1m 28s.

The build process encountered 5 warnings.

[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpCommandValueOfReusesKnownCommandsAndKeepsCustomNames() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderStreamsDataContentAndTerminatesLastContent() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpContentPreservesConcreteTypesAcrossCopyDuplicateAndReplace() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderRejectsLinesLongerThanConfiguredMaximum() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderWaitsForCompleteLinesAndRejectsMalformedLines() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderAccumulatesMultiLineResponsesAcrossInboundWrites() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > defaultMessagesValidateCodesAndImplementValueSemantics() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestFactoryCreatesProtocolSpecificCommandsAndImmutableParameters() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderParsesSingleLineAndMultiLineResponses() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines() FAILED


Failures (3):
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderStreamsDataContentAndTerminatesLastContent()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderStreamsDataContentAndTerminatesLastContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 32s

       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 8 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[        11 tests found           ]
[         0 tests skipped         ]
[        11 tests started         ]
[         0 tests aborted         ]
[         8 tests successful      ]
[         3 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-smtp:4.1.74.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 33s
11 actionable tasks: 3 executed, 8 up-to-date
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"]
```

## native-image run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: native-image run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25251908980/job/74045252624
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
/metadata/org.junit.platform/junit-platform-commons/1.8.2,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/org.hamcrest/hamcrest-core/1.3,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-codec-smtp/4.1.74.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-common/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-buffer/4.1.80.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-transport/4.1.115.Final,/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/metadata/io.netty/netty-resolver/4.1.79.Final \
-H:Optimize@user+api=b \
-H:+ReportExceptionStackTraces@user \
-H:FutureDefaults@user+api=all \
-H:Features@user+api=org.graalvm.junit.platform.JUnitPlatformFeature \
'-H:Class@explicit main-class=org.graalvm.junit.platform.NativeImageJUnitLauncher' \
-H:ImageBuildID@driver=116ce6f5-e1a5-b7f9-9795-9d18f2d1dd5b \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/thirdparty/native-image.properties+api=com.oracle.svm.thirdparty.gson.GsonFeature' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/library-support.jar!/META-INF/native-image/com.oracle.svm/polyglot/native-image.properties+api=com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature' \
'-H:AddExports@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=java.base/jdk.internal.misc=org.graalvm.nativeimage.guest' \
'-H:Features@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=com.oracle.svm.guest.features.jdk.hosted.JavaNetHttpFeature' \
'-H:ClassInitialization@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/native-image.properties+api=sun.net.httpserver.simpleserver:run_time' \
'-H:ReachabilityMetadataResources@jar:file:///opt/hostedtoolcache/graalvm-jdk-latest-ea_linux-x64_bin/25.0.0/x64/graalvm-25.1.0-dev+10.1/lib/svm/svm-guest.jar!/META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json+api=META-INF/native-image/org.graalvm.nativeimage.guest/reachability-metadata.json' \
-H:DriverWarningsCount=4
]
Warning: Option 'FallbackThreshold' is deprecated and might be removed in a future release: It no longer has any effect, and no replacement is available. Please refer to the GraalVM release notes.
========================================================================================================================
GraalVM Native Image: Generating 'io.netty.netty-codec-smtp_tests-tests' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                   (6.8s @ 0.21GiB)
 Builder configuration:
 - Java version: 25.0.2+10-LTS, vendor version: Oracle GraalVM 25.1.0-dev+10.1
 - Graal compiler: optimization level: b, target machine: x86-64-v3, PGO: off
 - C compiler: gcc (linux, x86_64, 13.3.0)
 - Assertions: disabled, system assertions: disabled
 - 1 user-specific feature(s):
   - org.graalvm.junit.platform.JUnitPlatformFeature
 Image configuration:
 - Garbage collector: Serial GC (max heap size: 80% of RAM)
 - Assertions: disabled (class-specific config may apply), system assertions: disabled
------------------------------------------------------------------------------------------------------------------------
 1 experimental option(s) unlocked:
 - '-H:BuildOutputJSONFile' (origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 13.27GiB of memory (85% of system memory because $CI set to 'true')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
[2/8] Performing analysis...  [*****]                                                                  (36.1s @ 0.97GiB)
    9,238 types,  10,677 fields, and  41,919 methods found reachable
      479 types,     118 fields, and     971 methods registered for reflection
       60 types,      68 fields, and      52 methods registered for JNI access
      153 resource accesses registered with 262.32KiB total size
        4 native libraries: dl, pthread, rt, z
[3/8] Building universe...                                                                              (5.4s @ 0.81GiB)
[4/8] Parsing methods...      [**]                                                                      (3.8s @ 0.89GiB)
[5/8] Inlining methods...     [***]                                                                     (2.1s @ 0.93GiB)
[6/8] Compiling methods...    [*****]                                                                  (21.0s @ 0.85GiB)
[7/8] Laying out methods...   [**]                                                                      (3.2s @ 1.04GiB)
[8/8] Creating image...       [**]                                                                      (3.9s @ 1.18GiB)
   9.86MiB (33.16%) for code area:    26,441 compilation units
  15.75MiB (52.95%) for image heap:  260,123 objects and 23 resources
   4.13MiB (13.89%) for other data
  29.74MiB in total image size, 27.50MiB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
   1.10MiB o.g.n.runtime/com.oracle.svm.core.code              3.09MiB byte[] for code metadata
 707.58KiB java.base/java.util                                 2.91MiB byte[] for string data
 408.42KiB java.base/jdk.internal.classfile.impl               1.54MiB java.base/java.lang.String
 315.39KiB java.base/java.lang                                 1.11MiB java.base/java.lang.Class
 267.01KiB java.base/com.sun.crypto.provider                   0.99MiB o.g.n.runtime/c.o.s.core.hub.DynamicHubCompanion
 231.38KiB java.base/java.text                               927.08KiB byte[] for general heap data
 230.65KiB io.netty.buffer                                   561.63KiB byte[] for reflection metadata
 226.75KiB java.base/java.util.concurrent                    525.20KiB heap alignment
 223.93KiB org.assertj.core.internal.bytebuddy.jar.asm       362.20KiB java.base/java.lang.Object[]
 192.84KiB java.base/sun.security.provider                   358.24KiB o.g.n.runtime/c.o.s.c.h.ImageReflectionMetadata
   6.03MiB for 336 more packages                               3.44MiB for 2207 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 22 component(s) is embedded in binary (1.13KiB). 1 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                       6.1s (7.1% of total time) in 821 GCs | Peak RSS: 1.80GiB | CPU load: 3.63
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests (executable, 27.50MiB)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library, 32.23KiB)
 /tmp/native-image-build-output.json (build_info, 1.03KiB)
========================================================================================================================
Finished generating 'io.netty.netty-codec-smtp_tests-tests' in 1m 24s.

The build process encountered 5 warnings.

[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpCommandValueOfReusesKnownCommandsAndKeepsCustomNames() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderStreamsDataContentAndTerminatesLastContent() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > smtpContentPreservesConcreteTypesAcrossCopyDuplicateAndReplace() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderRejectsLinesLongerThanConfiguredMaximum() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderWaitsForCompleteLinesAndRejectsMalformedLines() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderAccumulatesMultiLineResponsesAcrossInboundWrites() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > defaultMessagesValidateCodesAndImplementValueSemantics() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody() FAILED

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestFactoryCreatesProtocolSpecificCommandsAndImmutableParameters() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > responseDecoderParsesSingleLineAndMultiLineResponses() SUCCESSFUL

io_netty.netty_codec_smtp.Netty_codec_smtpTest > requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines() FAILED


Failures (3):
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderStreamsDataContentAndTerminatesLastContent()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderStreamsDataContentAndTerminatesLastContent', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderEnforcesDataStateAndAllowsRsetToAbortMailBody', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]
  JUnit Jupiter:Netty_codec_smtpTest:requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines()
    MethodSource [className = 'io_netty.netty_codec_smtp.Netty_codec_smtpTest', methodName = 'requestEncoderWritesCommandsWithArgumentsAndEmptyCommandLines', methodParameterTypes = '']
    => io.netty.handler.codec.EncoderException: java.lang.NullPointerException
       io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:830)
       io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:749)
       io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1031)
       io.netty.channel.Channel.write(Channel.java:272)
       [...]
     Caused by: java.lang.NullPointerException
       io.netty.util.internal.PlatformDependent0.safeConstructPutInt(PlatformDependent0.java:699)
       io.netty.util.internal.PlatformDependent.safeConstructPutInt(PlatformDependent.java:516)
       io.netty.util.internal.ReferenceCountUpdater.setInitialValue(ReferenceCountUpdater.java:67)
       io.netty.buffer.AdaptivePoolingAllocator$Chunk.<init>(AdaptivePoolingAllocator.java:840)
       io.netty.buffer.AdaptivePoolingAllocator$Magazine.newChunkAllocation(AdaptivePoolingAllocator.java:761)
       [...]

Test run finished after 5 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[        11 tests found           ]
[         0 tests skipped         ]
[        11 tests started         ]
[         0 tests aborted         ]
[         8 tests successful      ]
[         3 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
17 actionable tasks: 7 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build/native/nativeTestCompile/io.netty.netty-codec-smtp_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 28s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for io.netty:netty-codec-smtp:4.1.74.Final failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 30s
11 actionable tasks: 3 executed, 8 up-to-date
❌ ERROR: 'native-image run' for 4.2.1.Final failed with exit code 1.
FAILED[native-image run][4.2.1.Final][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.2.1.Final" ./gradlew clean test -Pcoordinates="io.netty:netty-codec-smtp:4.1.74.Final"]
```

--- a/metadata/io.netty/netty-codec-smtp/4.2.1.Final/reachability-metadata.json
+++ b/metadata/io.netty/netty-codec-smtp/4.2.1.Final/reachability-metadata.json
@@ -1,0 +1,197 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpRequestEncoder"
+      },
+      "type": "io.netty.buffer.AbstractByteBufAllocator"
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpRequestEncoder"
+      },
+      "type": "io.netty.buffer.AbstractReferenceCountedByteBuf",
+      "fields": [
+        {
+          "name": "refCnt"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpRequestEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueConsumerIndexField",
+      "fields": [
+        {
+          "name": "consumerIndex"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpRequestEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerIndexField",
+      "fields": [
+        {
+          "name": "producerIndex"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpRequestEncoder"
+      },
+      "type": "io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerLimitField",
+      "fields": [
+        {
+          "name": "producerLimit"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.lang.Thread",
+      "methods": [
+        {
+          "name": "isVirtual",
+          "parameterTypes": []
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.lang.management.ManagementFactory",
+      "methods": [
+        {
+          "name": "getRuntimeMXBean",
+          "parameterTypes": []
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.lang.management.RuntimeMXBean",
+      "methods": [
+        {
+          "name": "getInputArguments",
+          "parameterTypes": []
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.nio.Bits"
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.nio.Buffer",
+      "fields": [
+        {
+          "name": "address"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.nio.ByteBuffer",
+      "methods": [
+        {
+          "name": "alignedSlice",
+          "parameterTypes": [
+            "int"
+          ]
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "java.nio.DirectByteBuffer"
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "jdk.internal.misc.Unsafe",
+      "methods": [
+        {
+          "name": "getUnsafe",
+          "parameterTypes": []
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "sun.management.VMManagementImpl",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "compTimeMonitoringSupport"
+        },
+        {
+          "name": "currentThreadCpuTimeSupport"
+        },
+        {
+          "name": "objectMonitorUsageSupport"
+        },
+        {
+          "name": "otherThreadCpuTimeSupport"
+        },
+        {
+          "name": "remoteDiagnosticCommandsSupport"
+        },
+        {
+          "name": "synchronizerUsageSupport"
+        },
+        {
+          "name": "threadAllocatedMemorySupport"
+        },
+        {
+          "name": "threadContentionMonitoringSupport"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "sun.misc.Unsafe",
+      "fields": [
+        {
+          "name": "theUnsafe"
+        }
+      ],
+      "methods": [
+        {
+          "name": "invokeCleaner",
+          "parameterTypes": [
+            "java.nio.ByteBuffer"
+          ]
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "io.netty.handler.codec.smtp.SmtpCommand"
+      },
+      "type": "sun.misc.VM"
+    }
+  ]
+}

--- a/metadata/io.netty/netty-codec-smtp/index.json
+++ b/metadata/io.netty/netty-codec-smtp/index.json
@@ -1,6 +1,21 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "4.2.1.Final",
+    "test-version" : "4.1.74.Final",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-smtp/$version$/netty-codec-smtp-$version$-sources.jar",
+    "repository-url" : "https://github.com/netty/netty",
+    "test-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-smtp/$version$/netty-codec-smtp-$version$-test-sources.jar",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-smtp/$version$/netty-codec-smtp-$version$-javadoc.jar",
+    "description" : "Netty Codec SMTP provides SMTP protocol support for Netty by modeling SMTP commands, requests, responses, and message content. It includes an SMTP request encoder and response decoder for implementing asynchronous SMTP clients on Netty's channel pipeline.",
+    "tested-versions" : [
+      "4.2.1.Final"
+    ],
+    "allowed-packages" : [
+      "io.netty.handler.codec.smtp"
+    ]
+  },
+  {
     "metadata-version" : "4.1.74.Final",
     "source-code-url" : "https://repo.maven.apache.org/maven2/io/netty/netty-codec-smtp/$version$/netty-codec-smtp-$version$-sources.jar",
     "repository-url" : "https://github.com/netty/netty",

--- a/stats/io.netty/netty-codec-smtp/4.2.1.Final/stats.json
+++ b/stats/io.netty/netty-codec-smtp/4.2.1.Final/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "4.2.1.Final",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 979,
+        "missed" : 181,
+        "ratio" : 0.843966,
+        "total" : 1160
+      },
+      "line" : {
+        "covered" : 210,
+        "missed" : 42,
+        "ratio" : 0.833333,
+        "total" : 252
+      },
+      "method" : {
+        "covered" : 62,
+        "missed" : 21,
+        "ratio" : 0.746988,
+        "total" : 83
+      }
+    }
+  } ]
+}

--- a/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build.gradle
+++ b/tests/src/io.netty/netty-codec-smtp/4.1.74.Final/build.gradle
@@ -15,6 +15,16 @@ dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
 }
 
+tasks.named("nativeTest") {
+    if (libraryVersion == '4.2.1.Final') {
+        // Netty 4.2.x requires explicit unsafe access on Java 25 native runs.
+        runtimeArgs.addAll([
+                '-Dsun.misc.unsafe.memory.access=allow',
+                '-Dio.netty.tryReflectionSetAccessible=true'
+        ])
+    }
+}
+
 graalvmNative {
     agent {
         defaultMode = "conditional"

